testng-core/src/main/java/org/testng/internal/ClassImpl.java

```diff
@@ -118,7 +123,11 @@ private Object getDefaultInstance(boolean create, String errMsgPrefix) {
       if (create) {
         DetailedAttributes ea = newDetailedAttributes(create, errorMsgPrefix);
         CreationAttributes attributes = new CreationAttributes(m_testContext, null, ea);
-        result = new Object[] {Dispenser.newInstance(m_objectFactory).dispense(attributes)};
+        result =
+            new Object[] {
+              Dispenser.newInstance(m_testContext.getSuite().getObjectFactory())
```

Comment:
Why not use the same logic here? https://github.com/cbeust/testng/pull/2677/files#diff-73ef03ca958970a8d4a3f3c0764700a861f31f92a10d0b46930671bee51dfba8R98-R102

Maybe a private method could help.

Reply:
@juherr - I didnt understand this comment and the link you shared is taking me to the same class. Can you please help me understand ?

Reply:
Sorry, I was not explicit enough.
You have 2 different ways to create a dispenser in the same class. Is it expected? If it is expected, could you add a comment that explains why? If it is not expected, you can share a same builder method.

Reply:
@juherr - Yes you are right. One place is for the JUnit flow but I dont even see that codepath being executed and so I dont know why it exists even. I was first thinking that, this particular code path is going to get executed when we have `junit` attribute in suite set to `true` and I am running a bunch of `JUnit` tests using TestNG. But I noticed that the tests don't even take that path. It may be dead code but I cannot remove it until I know for sure. I will dig into it deeper separately and not in this PR.

Reply:
From looking at the code, it looks like we perhaps never supported the notion of a test object factory for JUnit tests, because the codepath (when I execute a JUnit annotated test using TestNG) is going such that, JUnit is internally invoking the constructor of the test class and our objectfactory is not even coming into picture.  I think if there are no other comments on this PR (which is to fix test objectfactory to be set for TestNG tests via listeners) we should merge this PR and deal with cleaning up the test object factory impact for JUnit tests separately (and maybe even remove the irrelevant code sections. )

Maybe I dont know enough but this is my first impression from what I see

Reply:
If you ask me, I will propose to drop the junit support and advice to use https://github.com/junit-team/testng-engine when users ask for running both junit and testng.

Reply:
@juherr sure. So can we go ahead with merging this PR ?

Reply:
just add a comment that link to the current discussion ;)

Reply:
I went ahead and captured this as a defect and included all the details https://github.com/cbeust/testng/issues/2683